### PR TITLE
refactor streamstore for single snapshot

### PR DESCRIPTION
**Pull request by ryanfkeepers**
Refactors the streamstore interface away from
multiple store clients, and towards a single client
that can collect multiple marshallable sets of
data.  This allows the streamstore to write all
entries into a single snapshot, instead of writing
N snapshots, one per each client.

---

#### Does this PR need a docs update or release note?

- [x] :no_entry: No

#### Type of change

- [x] :sunflower: Feature

#### Issue(s)

* #2708

#### Test Plan

- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

File: src/internal/streamstore/streamstore.go [hunk 1]
```diff
@@ -18,24 +18,102 @@ import (
...
+	service path.ServiceType,
+) *storeStreamer {
+	return &storeStreamer{
+		kw:      kw,
+		tenant:  tenant,
+		service: service,
+	}
+}
+
+func (ss *storeStreamer) Collect(ctx context.Context, col Collectable) error {
```
Comment (ashmrtn) [inline src/internal/streamstore/streamstore.go:48]: This needs a comment that the data is eagerly serialized

File: src/pkg/backup/backup.go [hunk 2]
```diff
@@ -22,13 +22,9 @@ type Backup struct {
...
-	// Reference to the backup details storage location.
-	// Used to read backup.Details from the streamstore.
-	DetailsID string `json:"detailsID"`
-
-	// Reference to the fault errors storage location.
-	// Used to read fault.Errors from the streamstore.
-	ErrorsID string `json:"errorsID"`
+	// Reference to the details and fault errors storage location.
+	// Used to read backup.Details and fault.Errors from the streamstore.
+	StreamStoreID string `json:"streamStoreID"`
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/pkg/backup/backup.go:27]: We still retain the detailsID, it's not lost.  Old backups should fall back to the detailsID if streamStoreID isn't found.  That'll be sufficient for lookups.

File: src/internal/streamstore/collectables.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	faultErrorsItemName = "fault_error"
+	faultErrorsPurpose  = "fault_error"
+
+	DetailsType     = "details"
+	detailsItemName = "details"
+	detailsPurpose  = "details"
+)
+
+// FaultErrorsCollector generates a collection of fault.Errors
+// containing the marshalled bytes from the provieded marshaller.
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/streamstore/collectables.go:22]: good caetch

File: src/internal/streamstore/collectables_test.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	}
+}
+
+func (suite *StreamStoreIntgSuite) TestStreamer() {
+	table := []struct {
+		name  string
+		deets func() *details.Details
+		errs  func() *fault.Errors
+	}{
+		{
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/streamstore/collectables_test.go:71]: agh, doh.  I really meant to have those.  Turned into a squirrel and chased other things instead.

File: src/internal/streamstore/mock/mock.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	case streamstore.DetailsType:
+		mr = ms.Deets[snapshotID]
+	case streamstore.FaultErrorsType:
+		mr = ms.Errors[snapshotID]
+	default:
+		return clues.New("unknown type: " + col.Type)
+	}
+
+	if mr == nil {
+		return errors.Errorf("no martshaller for %s ID %s", col.Type, snapshotID)
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/streamstore/mock/mock.go:45]: t strikes again!

File: src/internal/streamstore/streamstore.go (hunk 1 above)
Comment (ryanfkeepers) [reply src/internal/streamstore/streamstore.go:48]: done

File: src/internal/operations/backup.go [hunk 6]
```diff
@@ -668,19 +666,24 @@ func (op *BackupOperation) createBackupModels(
...
 	}
 
-	errorsID, err := errorsStore.Write(ctx, errs.Errors(), errs)
+	err = sscw.Collect(ctx, streamstore.FaultErrorsCollector(errs.Errors()))
 	if err != nil {
 		return clues.Wrap(err, "creating errors persistence").WithClues(ctx)
 	}
 
-	ctx = clues.Add(ctx, "details_id", detailsID)
+	ssid, err := sscw.Write(ctx, errs)
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/operations/backup.go:679]: It wasn't lost.  I didn't agree with the concern, is all.  We're not serializing the bus.  We're serializing a snapshotted production of the bus.  The two states can't interact, and I don't see any reason for us to change that behavior in the future to re-introduce that mutation; kind of an anti-pattern at large.

That said, I don't have strong feelings here, so I've added a separate fault bus for the persistence action.  Should be very safe now.

File: src/pkg/backup/backup.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/pkg/backup/backup.go:27]: for proof, this backup was generated prior to the creation of this branch:
```
make details service=exchange backup=91166111-a1a6-4336-b105-5f3e91cc3d2f
cd ./corso/src; go run ./corso.go backup details exchange \
	--config-file /Users/keepers/go/src/github.com/alcionai/.corso_make.toml --no-stats --log-file /Users/keepers/go/src/github.com/alcionai/.corso.log --log-level warn  \
	--backup 91166111-a1a6-4336-b105-5f3e91cc3d2f \
	
Logging to file: /Users/keepers/go/src/github.com/alcionai/.corso.log
Connecting to repository:      3s done
  ID            Contact Name
  8c23d44a13c2  Vaib Karma
  78440cf1caba  Ash Marks
```

File: src/internal/streamstore/mock/mock.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+var _ streamstore.Streamer = &Streamer{}
+
+type Streamer struct {
+	Deets  map[string]*details.Details
+	Errors map[string]*fault.Errors
+}
+
+func (ms Streamer) Collect(context.Context, streamstore.Collectable) error {
+	return clues.New("not implmented")
```
Comment (ashmrtn) [inline src/internal/streamstore/mock/mock.go:24]:
```suggestion
	return clues.New("not implemented")
```

File: src/internal/streamstore/mock/mock.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	bs, err := mr.Marshal()
+	if err != nil {
+		return err
+	}
+
+	return col.Unmr(io.NopCloser(bytes.NewReader(bs)))
+}
+
+func (ms Streamer) Write(context.Context, *fault.Bus) (string, error) {
+	return "", clues.New("not implmented")
```
Comment (ashmrtn) [inline src/internal/streamstore/mock/mock.go:57]:
```suggestion
	return "", clues.New("not implemented")
```

File: src/internal/streamstore/mock/mock.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+func (ms Streamer) Delete(context.Context, string) error {
+	return clues.New("not implmented")
```
Comment (ashmrtn) [inline src/internal/streamstore/mock/mock.go:61]:
```suggestion
	return clues.New("not implemented")
```

**Review by ashmrtn — APPROVED** [review]

**Pull request merged**